cluster_management/src/main/java/com/pinterest/rocksplicator/Participant.java

```diff
@@ -444,4 +369,209 @@ public void run() {
       });
     }
   }
+
+  /**
+   * This method is called from CPP code, when the service is about to do down.
+   * This will cause all states to transition to offline eventually.
+   */
+  public static void shutDownParticipant() throws Exception {
+
+    /**
+     * TODO: grajpurohit: Improve the logic to cleanly shutdown.
+     *
+     * This is probably not the right way. When a participant is going down,
+     * we need a chance to cleanly offline the currently top-state and second
+     * top-state resources (e.g.. all partitions state should be brought to offline
+     * state) before going down. That will be a graceful shutdown. However, when
+     * helixManager disconnects, it releases all message handlers / listeners there after
+     * there is no way for resources to be brought down through state transition.
+     *
+     * Ideal way will be whereby
+     *
+     * 1. We setup instanceTag "disabled" so that spectator has a first chance to be notified that
+     * this participant is going down without bringing down the hosted partition replicas.
+     *
+     * 2. Wait to ensure that the instanceTag has been added to the participant instance.
+     *
+     * 3. We should then disable this instance (currently this is the only way we figured out to
+     * force controller to issue state transitions to existing current states of all partitions
+     * hosted by this participant) to move to OFFLINE state.
+     *
+     * 4. Wait until all partition replicas hosted by this participant to have moved to OFFLINE state.
+     *
+     * 5. Now we are free to disconnect to HelixManager, so that all listeners and message callback
+     * handlers are properly un-registered.
+     */
+    if (helixManager != null && helixManager.isConnected()) {
+      disableInstance(helixManager.getClusterManagmentTool(), staticClusterName, staticInstanceId);
```

Comment:
should we otherwise,

disable instance
add disabled instance tag

wait for resources to be offlined
unregister the statemodelfactoriies
disconnect from helixManager.

//use separate helixAdmin, not one from helixManager above.
re-enable the instance (to ensure that when it comes up, the instance is already enabled)
remove the disabled instance-tag.